hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/replication/VerifyReplication.java

```diff
@@ -329,6 +343,17 @@ protected void cleanup(Context context) {
           LOG.error("fail to close replicated connection in cleanup", e);
         }
       }
+      if (reCompareExecutor != null && !reCompareExecutor.isShutdown()) {
```

Comment:
when we get to this point, there may be outstanding recompares. They may not finish within the 10s below. I think we need to fail the job if this happens, or mark all of those outstanding recompares as BADROWS. Otherwise, the job could complete successfully after having skipped over some recompares which. may have been bad. This would give the user a false sense of security.

Reply:
in order to solve this, you might need to make the SynchronousQueue available for introspection. So when you shutdown the executor, hopefully any running runnables will get interrupted (need to verify), and any that had been pending in the queue should be iterated and failed. You might want to add a `fail()` method to the RecompareRunnable which increments BADROWS, FAILED_RECOMPARE, and originalCounter.

Reply:
I think we also need to fail the currently executing runnable as well, which will no longer be in the queue afaict.

Reply:
so maybe something like this 

```java
int activelyExecutingTasks = reCompareExecutor.getActiveCount();
context.getCounter(Counters.FAILED_RECOMPARE).increment(activelyExecutingTasks);
context.getCounter(Counters.BADROWS).increment(activelyExecutingTasks);
```

although this doesn't let us increment the original counter, unfortunately. To do that I think we might need to manually track any active runnables

Reply:
i dont think getActiveCount is really thread safe. It's an approximation of the current number of running threads at the time, and doesn't include queued tasks. Given the importance of this job for identifying replication inconsistencies, i think we need a perfectly accurate count (or at least err on the side of overcounting failures, though that could be annoying)

So here's a simpler idea -- You could add an AtomicInteger, which gets incremented whenever a RecompareRunnable is created and then decremented when it completes. You'd increment these counters by whatever the value of that AtomicInteger is at the time.

The downside with that is as you said, we lose the original counter. I suppose we can live without it, though I think there are more complicated solutions which could ensure we capture those if we desired.

Reply:
summarizing the current solution:

1. We try to shutdown the executor and await termination. If the executor was terminated, we can return as normal
2. We call shutdownNow to interrupt any active threads and get the list of runnables that were waiting execution. Any runnables that were sleeping on recompares will fail and increment the correct counters
3. We call `fail()` on any runnables that were queued which will increment the BADROWS, FAILED_RECOMPARE and original counters
4. We await termination again to let any running deployables that weren't interrupted
5. If we still haven't terminated, increment BADROWS and FAILED_RECOMPARE `Math.max(1, activeCount)`. We're assuming any of the recompares would've failed to err on the safe side